game/audio.cc

```diff
@@ -62,173 +67,191 @@ namespace {
+// 			} else
+// #endif
+		if (t.mpeg.audioQueue(&*mixbuf.begin(), &*mixbuf.end(), m_pos, t.fadeLevel)) eof = false;
+		
 	}
-	/// Prepare (seek) all tracks to current position, return true when done (nonblocking)
-	bool prepare() {
-		bool ready = true;
-		for (auto& kv: tracks) {
-			FFmpeg& mpeg = kv.second->mpeg;
-			if (mpeg.terminating()) continue;  // Song loading failed or other error, won't ever get ready
-			if (mpeg.audioQueue.prepare(m_pos)) continue;  // Buffering done
-			ready = false;  // Need to wait for buffering
-			break;
+	m_pos += samples;
+	
+	for (size_t i = 0, iend = mixbuf.size(); i != iend; ++i) {
+		if (i % 2 == 0) {
+			fadeLevel += fadeRate;
+			if (fadeLevel <= 0.0) return false;
+			if (fadeLevel > 1.0) { fadeLevel = 1.0; fadeRate = 0.0; }
 		}
-		return ready;
+		begin[i] += mixbuf[i] * fadeLevel * static_cast<float>(m_preview ? config["audio/preview_volume"].i() : config["audio/music_volume"].i())/100.0;
 	}
-	void trackFade(std::string const& name, double fadeLevel) {
-		auto it = tracks.find(name);
-		if (it == tracks.end()) return;
-		it->second->fadeLevel = fadeLevel;
+	// suppress center channel vocals
+	if(suppressCenterChannel && !m_preview) {
+		float diffLR;
+		for (size_t i=0; i<mixbuf.size(); i+=2) {
+			diffLR = begin[i] - begin[i+1];
+			begin[i] = diffLR;
+			begin[i+1] = diffLR;
+		}
 	}
-	void trackPitchBend(std::string const& name, double pitchFactor) {
-		auto it = tracks.find(name);
-		if (it == tracks.end()) return;
-		it->second->pitchFactor = pitchFactor;
+	return !eof;
+}
+
+double Music::duration() const {
+	double dur = 0.0;
+	for (auto& kv: tracks) dur = std::max(dur, kv.second->mpeg.audioQueue.duration());
+	return dur;
+}
+
+bool Music::prepare() {
+	bool ready = true;
+	for (auto& kv: tracks) {
+		FFmpeg& mpeg = kv.second->mpeg;
+		if (mpeg.terminating()) continue;  // Song loading failed or other error, won't ever get ready
+		if (mpeg.audioQueue.prepare(m_pos)) {
+			if (kv.first == "background" && m_preview && m_pos > 0) {
+				fvec_t* previewSamples = mpeg.audioQueue.makePreviewBuffer();
+				fvec_t* previewBeats = ScreenSongs::previewBeatsBuffer.get();
+				intptr_t readptr = 0;
+				fvec_t* tempoSamplePtr = new_fvec(Audio::aubio_hop_size);
+				std::lock_guard<std::recursive_mutex> l(Music::aubio_mutex);
+// 				std::clog << "audio/debug: file: " << mpeg.getFilename().filename().string() << std::endl;
+				Game* gm = Game::getSingletonPtr();
+				ScreenSongs* sSongs = static_cast<ScreenSongs *>(gm->getScreen("Songs"));
+				double pstart = sSongs->getSongs().currentPtr()->preview_start;
+				pstart = (std::isnan(pstart) ? 0.0 : pstart);
+				double first_period, first_beat;
+				std::vector<double> extra_beats;
+				Song::Beats& beats = sSongs->getSongs().currentPtr()->beats;
+				if (!sSongs->getSongs().currentPtr()->hasControllers()) {
+				while (readptr < previewSamples->length) {
+// 						std::clog << "audio/debug: readptr: " << readptr << std::endl;
```

Comment:
Yeah I can remove it, but I'd left it in for same reason as above, on one hand it might be useful at some point (though I guess I could just write it again) but I didn't want more spam in the log.